Generate OSCORE object and use it for endpoint creation

rikard-sics opened the pull request:
Code to implement the points listed here:
https://github.com/eclipse/leshan/pull/718#issuecomment-521316754
- create the OSCORE object
- use it in leshan client demo (keeping hardcoded value but create oscore object)
- modify ServersInfoExtractor and ServerInfo to support oscore.
- use ServerInfo in CaliforniumEndpointsManager to create endpoint with oscore support.

The OSCORE object is generated and filled in the Leshan client demo. It will then be used to create a ServerInfo object using the ServersInfoExtractor. Next the ServerInfo is used in the CaliforniumEndpointsManager to generate endpoints towards servers that OSCORE should be used with.

See also issue regarding OSCORE Leshan client integration:
https://github.com/eclipse/leshan/issues/726

sbernard31 commented on leshan-client-core/pom.xml:
```diff
@@ -34,6 +35,12 @@ Contributors:
+                <groupId>org.eclipse.californium</groupId>
+                <artifactId>cf-oscore</artifactId>
+                <version>${californium.version}</version>
+        </dependency>
```
`leshan-client-code` must not depend to californium. (see https://github.com/eclipse/leshan/wiki/Modules#leshan-client-core)

rikard-sics replied:
I see, thanks for the link. I will amend the code so leshan-client-core does not have any dependencies from Californium. I can probably keep it as an int instead of AlgorithmID.

rikard-sics replied:
This should now be taken care of in commit 8fc7bfcdb67d219f712c5f2a1536a2d4c692e860
I needed the hex2ByteArray method from StringUtil in Californium. For now I extracted and took the code from there. If that is not fine I can re-implement it.

sbernard31 replied:
maybe you could find what you need in `org.eclipse.leshan.util.Hex` ?

rikard-sics replied:
> maybe you could find what you need in `org.eclipse.leshan.util.Hex` ?

Yes, I will check that.

rikard-sics replied:
Update to use a method from the Leshan Hex class rather than the imported method from Californium in commit ebc728a69129645fe1347e312c1feb34ee4aae13

sbernard31 commented on leshan-client-demo/src/main/java/org/eclipse/leshan/client/demo/LeshanClientDemo.java:
```diff
@@ -398,6 +372,10 @@ public static void createAndStartClient(String endpoint, String localAddress, in
+            } else if (useOSCore) {
+                initializer.setInstancesForObject(SECURITY, secOSCore(serverURI, 123));
+                initializer.setInstancesForObject(OSCORE, Oscore.set("11223344", "AA", "BB")); //Hardcoded values
+                initializer.setInstancesForObject(SERVER, new Server(123, 30, BindingMode.U, false));
```
Regarding your remarks here : 
> One issue I had was how exactly to link the Security and Oscore objects together. I now see your suggestion here: #726 (comment)

To be able to do that, you will probably need to add a new XML version of [security object](http://www.openmobilealliance.org/release/LightweightM2M/V1_1_1-20190617-A/HTML-Version/OMA-TS-LightweightM2M_Core-V1_1_1-20190617-A.html#Table-E1-1-LwM2M-Object-LWM2M-Security-object-definition).

You should get it [here](http://www.openmobilealliance.org/wp/OMNA/LwM2M/LwM2MRegistry.html)

rikard-sics replied:
The security object xml has been updated and a link to the OSCORE object introduced in commit ecebdfa09ce8028fa02d0d26c448ce3029069479

sbernard31 commented on leshan-core/src/main/java/org/eclipse/leshan/SecurityMode.java:
```diff
@@ -12,14 +12,15 @@
  */
 public enum SecurityMode {
-    PSK(0), RPK(1), X509(2), NO_SEC(3);
+    PSK(0), RPK(1), X509(2), NO_SEC(3), OSCORE(4);
```
Regarding the [specification](http://www.openmobilealliance.org/release/LightweightM2M/V1_1_1-20190617-A/HTML-Version/OMA-TS-LightweightM2M_Core-V1_1_1-20190617-A.html#Table-E1-1-LwM2M-Object-LWM2M-Security-object-definition), 4 is for _"Certificate mode with EST"_ and there is nothing for OSCORE :thinking: .

(It is not excluded there is issue in the specification but for now, I don't understand all of this enough to say if this is an issue :grin:)

rikard-sics replied:
I now removed the SecurityMode for OSCORE. Instead the OSCORE Security Mode link in the Security object will be used. Done in commit 3a4e777eb664aa126c9c4bae1eb1effba7eace0e

sbernard31 commented on leshan-client-core/src/main/java/org/eclipse/leshan/client/servers/ServersInfoExtractor.java:
```diff
@@ -112,6 +122,16 @@ public static ServersInfo getInfo(Map<Integer, LwM2mObjectEnabler> objectEnabler
                         info.clientCertificate = getClientCertificate(security);
                         info.serverCertificate = getServerCertificate(security);
                         info.privateKey = getPrivateKey(security);
+                    } else if (info.secureMode == SecurityMode.OSCORE) {
```
as there is no OSCORE securityMode maybe we can just look if the resource OSCORE Security Mode(17) is set.

If this is set you get the right index for the FIXME below :)

rikard-sics replied:
Yes, I am no longer checking the security mode here to decide what to do but rather if the link from the Security object has been set. I am also using the link from the Security object to find the OSCORE object here. Done in commit 3e7db26f5131ad4ef4622ae864e69d5796b3eefa

sbernard31 commented on leshan-client-core/src/main/java/org/eclipse/leshan/client/object/Oscore.java:
```diff
@@ -32,78 +44,140 @@
+        int defaultHmacAlgorithm = -10; //HKDF_HMAC_SHA_256
+        
+        return new Oscore(masterSecret, senderId, recipientId, defaultAeadAlgorithm, defaultHmacAlgorithm, "");
+    }
```
What is a benefits of both static "constructor" above ?

sbernard31 replied:
You can create several constructors with different number of parameter.

(static constructor is a good way to name a constructor but the con is that you can not reused when you extend the class)

rikard-sics replied:
> You can create several constructors with different number of parameter.
> 
> (static constructor is a good way to name a constructor but the con is that you can not reused when you extend the class)

Okay, so it would be better to have 1 or several normal constructors, but no static constructors for the Oscore class?

sbernard31 replied:
Yep, as it seems you don't really need to "name" your constructor, I think it's better to have several "normal" constructors.

rikard-sics replied:
Removed the static "constructors" in commit dc37bd9c5a597cb41ca91238edbfe76665b72c76

sbernard31 commented on leshan-client-core/src/main/java/org/eclipse/leshan/client/object/Security.java:
```diff
@@ -108,6 +109,14 @@ public static Security noSec(String serverUri, int shortServerId) {
+    /**
+     * Returns a new security instance (OSCORE) for a device management server.
+     */
+    public static Security secOSCore(String serverUri, int shortServerId) {
```
here this would be more something like :
- oscore (or oscoreOnly)

and one day we will add something like this ? 
- pskWithOscore 
- rpkWithOscore
- x509WithOscore

rikard-sics replied:
Yes, that makes sense. I can rename the current method to oscore/oscoreOnly. And then we proceed with adding methods later that combine DTLS+OSCORE.

rikard-sics replied:
Renamed the current method to oscoreOnly in commit dd58e43e1da396aa35c54a24b898752647c04bdd . In the future further methods can be added.

sbernard31 commented on leshan-core-cf/src/main/java/org/eclipse/leshan/core/californium/EndpointFactory.java:
```diff
@@ -28,6 +30,8 @@
 
     CoapEndpoint createUnsecuredEndpoint(InetSocketAddress address, NetworkConfig coapConfig, ObservationStore store);
+    
+    CoapEndpoint createOSCoreEndpoint(InetSocketAddress address, NetworkConfig coapConfig, ObservationStore store, HashMapCtxDB db);
```
for now my feeling is that there no need to add createOSCoreEndpoint maybe just add the db parameter to both other methods (but as the californium API is not yet finished, I maybe missed something)

rikard-sics replied:
Yes I removed this new method and added the db parameter to the others instead. Done in commit 495d5019f750ffee544ce7a66bb27c270e372de7

rikard-sics commented:
> I suppose we should add a new attribute for OSCORE which could be used in addition of existing DTLS identity (in case of OSCORE over coaps) or without it (in case of OSCORE over coap).

Yes, that makes sense.

> In oscore what could be the value which could be used as an identifier ? I need to read the rfc again (intuitively I suppose that Sender ID or/and Recipient ID could be candidate?)

A combination of the sender ID, recipient ID and ID context should identify a pair of endpoints communicating using OSCORE.

sbernard31 commented on leshan-client-demo/src/main/java/org/eclipse/leshan/client/demo/LeshanClientDemo.java:
```diff
@@ -373,8 +373,10 @@ public static void createAndStartClient(String endpoint, String localAddress, in
-                initializer.setInstancesForObject(SECURITY, oscoreOnly(serverURI, 123));
-                initializer.setInstancesForObject(OSCORE, new Oscore("11223344", "AA", "BB")); //Hardcoded values
+            	Oscore oscoreObject = new Oscore("11223344", "AA", "BB"); //Hardcoded values
+            	oscoreObject.setId(12345);
```
if you want you can create an Oscore constructor which take an instanceID in parameter

rikard-sics replied:
Yes, I will do that instead.

rikard-sics replied:
Now done in commit ae73ab25b69513f1b63b8bf4e2c9fbc4ca931a50

sbernard31 commented:
You should have a look at [§5.5.5 - Endpoint Client Name](http://www.openmobilealliance.org/release/LightweightM2M/V1_1-20180710-A/HTML-Version/OMA-TS-LightweightM2M_Transport-V1_1-20180710-A.html#5-5-5-0-555-Endpoint-Client-Name) about you "identity" concern.

About that is sender ID / receiver ID available in californium endpoint context ?

sbernard31 commented on leshan-client-core/src/main/java/org/eclipse/leshan/client/object/Oscore.java:
```diff
@@ -64,8 +64,9 @@ public Oscore() {
-    public Oscore(String masterSecret, String senderId, String recipientId, int aeadAlgorithm, int hkdfAlgorithm, String masterSalt) {
-        this.masterSecret = masterSecret;
+    public Oscore(int instanceId, String masterSecret, String senderId, String recipientId, int aeadAlgorithm, int hkdfAlgorithm, String masterSalt) {
+        this.setId(instanceId);
```
clearly not a big deal but you can use `super(id)`

rikard-sics replied:
True, I will change that as part of the next commit.

sbernard31 commented on leshan-client-core/src/main/java/org/eclipse/leshan/client/servers/ServersInfoExtractor.java:
```diff
@@ -110,7 +110,34 @@ public static ServersInfo getInfo(Map<Integer, LwM2mObjectEnabler> objectEnabler
+                    boolean useOscore = false;
+                    if(oscoreObjectId == OSCORE) {
+                    	useOscore = true;
+                    }
```
boolean useOscore = oscoreObjectId == OSCORE ?

rikard-sics replied:
Yeah, looks more compact and nicer. I will change it.

sbernard31 commented on leshan-client-core/src/main/java/org/eclipse/leshan/client/servers/ServersInfoExtractor.java:
```diff
@@ -110,7 +110,34 @@ public static ServersInfo getInfo(Map<Integer, LwM2mObjectEnabler> objectEnabler
+                    if(useOscore) {
+                    	// search corresponding oscore object
+                    	LwM2mObjectInstance oscoreInstance = null;
+                        for (LwM2mObjectInstance oscore : oscores.getInstances().values()) {
```
Not need to iterate, there is a getter by Id => `oscores.getInstance(id)`

rikard-sics replied:
Yes, that is much nicer. Fixed in commit a255bd64d5334790597db3d9de6addcf81b78433

rikard-sics commented:
> About that is sender ID / receiver ID available in californium endpoint context ?

Yes it is available but only from Californium 2.0.0-M17. Since the context DB is a singleton in the Californium we are using now the values could also be gotten from there (for now).

rikard-sics commented:
> You should have a look at [§5.5.5 - Endpoint Client Name](http://www.openmobilealliance.org/release/LightweightM2M/V1_1-20180710-A/HTML-Version/OMA-TS-LightweightM2M_Transport-V1_1-20180710-A.html#5-5-5-0-555-Endpoint-Client-Name) about you "identity" concern.
> 

I now updated the code so that the Sender ID of the client is equal to its name (endpoint). So if I start a client with name "c1" its sender ID will be 0x6331, and on the server side in the web interface the sender ID must be entered as "6331" (hexadecimal). Done in commit 401db2bfd59b9481cc5f452a524b7c52bdb84f19

rikard-sics commented:
For further work on the identity I would perhaps wait until after updating to Californium 2.0.0-M17. There the sender ID, recipient ID and ID context are conveniently available in the endpoint context.

sbernard31 commented on leshan-client-cf/src/main/java/org/eclipse/leshan/client/californium/impl/CaliforniumEndpointsManager.java:
```diff
@@ -97,36 +97,9 @@ public synchronized Server createEndpoint(ServerInfo serverInfo) {
-                System.out.println("Adding OSCORE CTX " + serverInfo.getFullUri().toASCIIString());
+            // oscore only mode
+            if(serverInfo.useOscore) {
+            	System.out.println("Adding OSCORE CTX " + serverInfo.getFullUri().toASCIIString());
```
Please use LOG instead of sysout ? :)

rikard-sics replied:
Yes, that was an old debug statement I should have fixed. Now done in commit a83918e1619c160747895ccc6e44731d1d90a6ed

rikard-sics commented:
I had to do one more commit regarding the registration, Identity and SecurityChecker. Previously I could not register an OSCORE client using the name I configured it with in the web interface of the server. Since it counted as not secure it but yet had a non-null securityInfo it wanted me to use DTLS. I fixed this in commit 40213f6bfc1fae4dd4c0d51eb3d48d17a1cfbd5a

The SecurityChecker should be expanded with checks to the identity when using OSCORE. Also the Identity class expanded to support OSCORE-related identities. SecurityInfo already has a useOSCore flag.

sbernard31 commented on leshan-client-core/src/main/java/org/eclipse/leshan/client/servers/ServerInfo.java:
```diff
@@ -65,7 +66,7 @@ public URI getFullUri() {
 
     public boolean isSecure() {
-        return secureMode != SecurityMode.NO_SEC;
+        return secureMode != SecurityMode.NO_SEC || useOscore == true;
```
I know this is a bit strange but here I will not really change this.

Historically this was used to know if we should use unsecured(coap) endpoint or secured(coaps) endpoint.

So maybe the right name is "isCoaps" ?
But anyway I read the code again and maybe this does not make so much sense now, so it would be more wise to not touch it for now.

rikard-sics replied:
Yes, fair enough, I do not need to change the functionality of isSecure. I see your point that it is more about coap vs. coaps (and OSCORE could apply to both cases). I changed this in commit 17098a79932f4c48aa608b7418edc3119a193be5

sbernard31 commented:
About `CaliforniumEndpointsManager`, maybe I missed something but the idea is to begin by supporting Oscore without DTLS right ?

So I think the oscore stuff is not at "The Good Place".
![this is the bad place!](https://proxy.duckduckgo.com/iu/?u=https%3A%2F%2Fmedia1.tenor.com%2Fimages%2F32151ae908b613741e651571f35ef2c2%2Ftenor.gif%3Fitemid%3D10603393&f=1&nofb=1)

> For further work on the identity I would perhaps wait until after updating to Californium 2.0.0-M17. There the sender ID, recipient ID and ID context are conveniently available in the endpoint context.

makes sense !

rikard-sics commented:
> About `CaliforniumEndpointsManager`, maybe I missed something but the idea is to begin by supporting Oscore without DTLS right ?

Yes, my intention was to start by supporting OSCORE without DTLS. So the good place would be outside of the `if(serverInfo.isSecure())`, like down by row 220?

sbernard31 commented:
> So the good place would be outside of the if(serverInfo.isSecure()), like down by row 220?

yep :)

rikard-sics commented:
> yep :)

Okay, sounds good :)

I have now done this in commit 7116077e8e6ed4a1a8456b9040333ad01875b146

Perhaps at some later stage it could be extracted to a separate method since we may want to combine the OSCORE configuration of an endpoint with different DTLS modes in a convenient way and minimize code reuse.

sbernard31 commented:
> Perhaps at some later stage it could be extracted to a separate method since we may want to combine the OSCORE configuration of an endpoint with different DTLS modes in a convenient way and minimize code reuse.

Correct, we will see when we will add this.


I think we are good, do you plan to add more stuff in this PR ? (maybe just fix formatting issue?)

rikard-sics commented:
> I think we are good, do you plan to add more stuff in this PR ? (maybe just fix formatting issue?)

Yes, fixing the formatting was the last thing I had planned. I will proceed with that.

rikard-sics commented:
> I think we are good, do you plan to add more stuff in this PR ? (maybe just fix formatting issue?)

I have now applied the Leshan Eclipse formatter to the parts of the code I have changed. Done in commit 0138768e1b906ad8f81f67389103266832360d66

sbernard31 commented:
I squashed all your commits in one (e31012710a149b619852f6b741e8ea4da083f3e6) which is now integrated in oscore branch.

I tried a "mvn clean install" and that failed so I fixed issues, see commit cc6136a63f081cfa8f8280a6ef67ff53a84288ee
Do not hesitate to launch this often, tests is here to help you ;)

About formatting I find several more issues, see commit e2ae4bca7cffca3430876a732a793defebf1b22d.

Are you using eclipse ? Did you follow [this guide](https://github.com/eclipse/leshan/wiki/Code-&-design-guidelines#configure-your-ide). In theory "format on save" is activated per project, so it should be very difficult to write bad formatted code :thinking:.

What do you plan as next steps ? :)

rikard-sics commented:
> I squashed all your commits in one ([e310127](https://github.com/eclipse/leshan/commit/e31012710a149b619852f6b741e8ea4da083f3e6)) which is now integrated in oscore branch.

Thanks!

> I tried a "mvn clean install" and that failed so I fixed issues, see commit [cc6136a](https://github.com/eclipse/leshan/commit/cc6136a63f081cfa8f8280a6ef67ff53a84288ee)
> Do not hesitate to launch this often, tests is here to help you ;)

Yes, I was using Eclipse and not running the maven tests from the command line like that. I should have ran tests more frequently and will try to do so in the future.

> About formatting I find several more issues, see commit [e2ae4bc](https://github.com/eclipse/leshan/commit/e2ae4bca7cffca3430876a732a793defebf1b22d).
> 
> Are you using eclipse ? Did you follow [this guide](https://github.com/eclipse/leshan/wiki/Code-&-design-guidelines#configure-your-ide). In theory "format on save" is activated per project, so it should be very difficult to write bad formatted code .

Yes, I am using Eclipse. Unfortunately up until now I have not been using the "format on save" feature. But rather formatting manually with CTRL+SHIFT+F. I will start using "format on save" as you suggest.

> What do you plan as next steps ? :)

Some thoughts for next steps:

- Update code to Californium 2.0.0-M17
- Implement an Identity and related things in SecurityChecker for OSCORE
- Work on the server side issues according to discussion in https://github.com/eclipse/leshan/issues/727
- Add possibility of entering OSCORE context information as command line parameters for client

What do you think makes most sense to prioritize?

sbernard31 commented:
This order makes sense to me :)

rikard-sics commented:
> This order makes sense to me :)

Okay, sounds good. I will proceed accordingly. :)
